crates/project/src/search_history.rs

```diff
@@ -0,0 +1,258 @@
+/// Determines the behavior to use when inserting a new query into the search history.
+#[derive(Default, Debug, Clone, PartialEq)]
+pub enum QueryInsertionBehavior {
+    #[default]
+    /// Always insert the query to the search history.
+    AlwaysInsert,
+    /// Replace the previous query in the search history, if the new query contains the previous query.
+    ReplacePreviousIfContains,
+}
+
+/// A handle that maintains the current selection in the search history.
+/// This can be passed to the search history to update the selection accordingly,
+/// e.g. when using the up and down arrow keys to navigate the search history.
+#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
+pub struct SearchHistorySelectionHandle {
```

Comment:
I think we should somehow indicate that the `selection` can get stale, at least in the rustdoc.

And let's rename it to `SearchHistoryCursor` maybe? Then we don't have to use in rustdocs and explain what a "selection" is and see it as a pointer of a kind.